metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1123,29 +1165,104 @@ private void claim(int epoch) {
+        boolean zkMigrationEnabled,
+        BootstrapMetadata bootstrapMetadata,
+        FeatureControlManager featureControl,
+        Consumer<ApiMessageAndVersion> recordConsumer
+    ) {
+        if (isLogEmpty) {
+            // If no records have been replayed, we need to write out the bootstrap records.
+            // This will include the new metadata.version, as well as things like SCRAM
+            // initialization, etc.
+            log.info("The metadata log appears to be empty. Appending {} bootstrap record(s) " +
+                "at metadata.version {} from {}.", bootstrapMetadata.records().size(),
+                bootstrapMetadata.metadataVersion(), bootstrapMetadata.source());
+            bootstrapMetadata.records().forEach(recordConsumer);
+
+            if (bootstrapMetadata.metadataVersion().isMigrationSupported()) {
+                if (zkMigrationEnabled) {
+                    log.info("Putting the controller into pre-migration mode. No metadata updates will be allowed until " +
+                        "the ZK metadata has been migrated");
+                    recordConsumer.accept(ZkMigrationState.PRE_MIGRATION.toRecord());
+                } else {
+                    log.debug("Setting the ZK migration state to NONE since this is a de-novo KRaft cluster.");
+                    recordConsumer.accept(ZkMigrationState.NONE.toRecord());
+                }
+            } else {
+                if (zkMigrationEnabled) {
+                    throw new RuntimeException("The bootstrap metadata.version " + bootstrapMetadata.metadataVersion() +
+                        " does not support ZK migrations. Cannot continue with ZK migrations enabled.");
+                }
+            }
+        } else {
+            // Logs have been replayed. We need to initialize some things here if upgrading from older KRaft versions
+            if (featureControl.metadataVersion().equals(MetadataVersion.MINIMUM_KRAFT_VERSION)) {
+                log.info("No metadata.version feature level record was found in the log. " +
+                    "Treating the log as version {}.", MetadataVersion.MINIMUM_KRAFT_VERSION);
+            }
+
+            if (featureControl.metadataVersion().isMigrationSupported()) {
+                log.info("Loaded ZK migration state of {}", featureControl.zkMigrationState());
+                switch (featureControl.zkMigrationState()) {
+                    case NONE:
+                        // Since this is the default state there may or may not be an actual NONE in the log. Regardless,
+                        // it will eventually be persisted in a snapshot, so we don't need to explicitly write it here.
+                        if (zkMigrationEnabled) {
+                            throw new RuntimeException("Should not have ZK migrations enabled on a cluster that was created in KRaft mode.");
+                        }
+                        break;
+                    case PRE_MIGRATION:
```

Comment:
I don't think this is correct. We could have the following sequence:

1. controller 0 writes migration record setting PRE_MIGRATION state
2. controller 0 tries to load ZK image, fails for some reason (this is all in a metadata transaction)
3. fail over to controller 1

We don't want to fail at step 3 , right? It's valid to fail over when in PRE_MIGRATION state.

Reply:
(So in other words, I think this is worthy of a `warn` level log, but not an exception)